### PR TITLE
XML lifting & lowering transformations

### PR DESCRIPTION
**Pull request by cermakmarek**
Added XSLT lifting and lowering generators (no SPARQL yet).

### CODE REVIEW

File: packages/core/src/xml-transformations/xslt-lowering-writer.ts [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+  XmlMatch,
+} from "./xslt-model";
+
+import { XmlWriter, XmlStreamWriter } from "../xml-schema/xml-writer";
+
+import { XSLT_LOWERING } from "./xslt-vocabulary";
+
+const xslNamespace = "http://www.w3.org/1999/XSL/Transform";
+
+export async function saveXsltLoweringToDirectory(
```
Comment (skodapetr) [inline packages/core/src/xml-transformations/xslt-lowering-writer.ts:20]: I know we use to have this method, but we should remove it. The writers should not access FileSystem directly.

File: packages/core/src/xml-transformations/xslt-lowering-writer.ts [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+      model.targetNamespacePrefix,
+      model.targetNamespace
+    );
+  }
+}
+
+async function writeSettings(
+  writer: XmlWriter
+): Promise<void> {
+  await writer.writeElementFull("xsl", "output")(async writer => {
```
Comment (skodapetr) [inline packages/core/src/xml-transformations/xslt-lowering-writer.ts:89]: I see there is a lot of awaits here .. can't you move the await into writeElementFull ? As a result you would not need to type them here so extensively.

File: packages/core/src/xml-transformations/xslt-model-adapter.ts [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+const simpleTypeMap: Record<string, string> = {
+  [OFN.boolean]: XSD.boolean,
+  [OFN.date]: XSD.date,
+  [OFN.time]: XSD.time,
+  [OFN.dateTime]: XSD.dateTimeStamp,
+  [OFN.integer]: XSD.integer,
+  [OFN.decimal]: XSD.decimal,
+  [OFN.url]: XSD.anyURI,
+  [OFN.string]: XSD.string,
+};
```
Comment (skodapetr) [inline packages/core/src/xml-transformations/xslt-model-adapter.ts:56]: You have similar fragment of code in xml-schema-model-adapter:
```
const simpleTypeMap: Record<string, QName> = {
  [OFN.boolean]: ["xs", "boolean"],
  [OFN.date]: ["xs", "date"],
  [OFN.time]: ["xs", "time"],
  [OFN.dateTime]: ["xs", "dateTimeStamp"],
  [OFN.integer]: ["xs", "integer"],
  [OFN.decimal]: ["xs", "decimal"],
  [OFN.url]: ["xs", "anyURI"],
  [OFN.string]: ["xs", "string"],
  [OFN.text]: langStringName,
};````
Are the related? Is so they should be extracted to common place.

File: packages/core/src/xml-transformations/xslt-vocabulary.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export const XSLT_LIFTING = {
+  Generator: "http://example.com/generator/xslt-lifting",
```
Comment (skodapetr) [inline packages/core/src/xml-transformations/xslt-vocabulary.ts:2]: @jakubklimek  We should not be using example namespace any longer, also now that we have the demo we should have our own prefix right?

**Review by skodapetr — COMMENTED** [review]
You do not have to modify the code in this PR but we should be clear about the comments before the merge.

File: packages/core/src/generator/generator-factory.ts [hunk 5]
```diff
@@ -19,6 +20,8 @@ export function createDefaultGenerator(
     new BikeshedGenerator(),
     new JsonSchemaGenerator(),
     new XmlSchemaGenerator(),
+    new XsltGenerator(false),
+    new XsltGenerator(true),
```
[the remark this answers is not in the document]
Comment (cermakmarek) [reply packages/core/src/generator/generator-factory.ts]: I agree that is a better solution; I have changed it to be two separate classes (`XsltLiftingGenerator` and `XsltLoweringGenerator`). There is however still a lot of redundancy shared for all the generators (at least JSON, XSD, XSLT); I would welcome a base class which would be parametrized only by the respective model adapters (e.g. `structureModelToXslt`) and writers (e.g. `writeXsltLifting`/`writeXsltLowering`).

File: packages/core/src/xml-transformations/xslt-lowering-writer.ts (hunk 1 above)
Comment (cermakmarek) [reply packages/core/src/xml-transformations/xslt-lowering-writer.ts:20]: This function might not be needed anymore; I have copied it over from the XSD generator, so I can remove them from both generators eventually.

File: packages/core/src/xml-transformations/xslt-lowering-writer.ts (hunk 2 above)
Comment (cermakmarek) [reply packages/core/src/xml-transformations/xslt-lowering-writer.ts:89]: Perhaps; I could design a new writer interface with non-async methods which would buffer the calls and write that all at once at the end.

File: packages/core/src/xml-transformations/xslt-model-adapter.ts (hunk 3 above)
Comment (cermakmarek) [reply packages/core/src/xml-transformations/xslt-model-adapter.ts:56]: Yeah, this and couple other pieces will be extracted to a common module.

File: packages/core/src/generator/generator-factory.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (skodapetr) [reply packages/core/src/generator/generator-factory.ts]: Would you please create an issue or perhaps a discussion for this topic so we can address it later?

**Pull request merged**